ddl/ddl_api.go

```diff
@@ -606,7 +606,13 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef, o
 	// Set `NoDefaultValueFlag` if this field doesn't have a default value and
 	// it is `not null` and not an `AUTO_INCREMENT` field or `TIMESTAMP` field.
 	setNoDefaultValueFlag(col, hasDefaultValue)
-	if col.FieldType.EvalType().IsStringKind() && col.Charset == charset.CharsetBin {
+
+	// We need the MySQL compatibility flag, see # 13992.
+	isBinaryCollation := func(collation string) bool {
+		// @TODO need define a const for "_bin".
+		return strings.Count(collation, "_bin") == 1
```

Comment:
Try changing to "return collation == charset.CollationBin"?

Reply:
But I saw in the MySQL, if you set the collate like XXX_bin the flag always BINARY.

Reply:
I read the MySQL code and it feels like you said. I see some tests in CI fail, such as `testInferTypeSuite.TestInferType`. It needs to update this test. The test such as `testIntegrationSuite3.TestChangingCharsetToUtf8`. It seems the code we need to update in https://github.com/pingcap/tidb/blob/master/ddl/table.go#L684.
Other tests also need to be handled, if you encounter problems, you can @me.